File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/pigeon_implement/EpoManagerDelegateImpl.kt
package com.idosmart.pigeon_implement

import com.idosmart.pigeon_implement.IDOEpoManagerDelegate
import com.idosmart.pigeon_implement.IDOEpoUpgradeStatus
import com.idosmart.pigeongen.api_epo.ApiEpoManagerDelegate
import com.idosmart.pigeongen.api_epo.ApiEpoUpgradeStatus
import com.idosmart.pigeongen.api_epo.ApiOTAGpsInfo


internal class EpoManagerDelegateImpl : ApiEpoManagerDelegate {

    companion object {
        @Volatile
        private var instance: EpoManagerDelegateImpl? = null
        fun instance(): EpoManagerDelegateImpl {
            return instance ?: synchronized(this) {
                instance ?: EpoManagerDelegateImpl().also { instance = it }
            }
        }
    }

    private constructor()

    var delegateGps: IDOEpoManagerDelegate? = null
    var upgradeStatus: IDOEpoUpgradeStatus = IDOEpoUpgradeStatus.IDLE

    var callbackUpgradeStatus: ((IDOEpoUpgradeStatus) -> Unit)? = null
    var callbackDownProgress: ((Float) -> Unit)? = null
    var callbackSendProgress: ((Float) -> Unit)? = null
    var callbackComplete: ((Int) -> Unit)? = null

    override fun onEpoStatusChanged(status: ApiEpoUpgradeStatus) {
        upgradeStatus = IDOEpoUpgradeStatus.ofRaw(status.raw)!!
        callbackUpgradeStatus?.invoke(upgradeStatus)
    }

    override fun onEpoDownloadProgress(progress: Double) {
        callbackDownProgress?.invoke(progress.toFloat())
    }

    override fun onEpoSendProgress(progress: Double) {
        callbackSendProgress?.invoke(progress.toFloat())
    }

    override fun onEpoComplete(errorCode: Long) {
        callbackComplete?.invoke(errorCode.toInt())
    }

    override fun onGetGps(callback: (Result<ApiOTAGpsInfo?>) -> Unit) {
        val gpsInfo = delegateGps?.getAppGpsInfo()
        if (gpsInfo != null) {
            val rs = ApiOTAGpsInfo(
                tcxoOffset = 0,
                longitude = gpsInfo.longitude.toDouble(),
                latitude = gpsInfo.latitude.toDouble(),
                altitude = gpsInfo.altitude.toDouble()
            )
            callback(Result.success(rs))
        } else {
            callback(Result.success(null))
        }
    }


}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/pigeon_implement/EpoManagerDelegateImpl.swift
//
//  EpoManagerDelegate.swift
//  protocol_channel
//
//  Created by hc on 2024/8/8.
//

import Foundation

class EpoManagerDelegateImpl: ApiEpoManagerDelegate {
    static let shared = EpoManagerDelegateImpl()
    private init() {}
    
    weak var delegateGps: IDOEpoManagerDelegate?
    private(set) var upgradeStatus: IDOEpoUpgradeStatus = .idle
    
    var callbackUpgradeStatus: ((IDOEpoUpgradeStatus) -> Void)?
    var callbackDownProgress: ((Float) -> Void)?
    var callbackSendProgress: ((Float) -> Void)?
    var callbackComplete: ((Int) -> Void)?
    
    func onEpoStatusChanged(status: ApiEpoUpgradeStatus) throws {
        upgradeStatus = IDOEpoUpgradeStatus(rawValue: status.rawValue)!
        callbackUpgradeStatus?(upgradeStatus)
    }
    
    func onEpoDownloadProgress(progress: Double) throws {
        callbackDownProgress?(Float(progress))
    }
    
    func onEpoSendProgress(progress: Double) throws {
        callbackSendProgress?(Float(progress))
    }
    
    func onEpoComplete(errorCode: Int64) throws {
        callbackComplete?(errorCode.int)
    }
    
    func onGetGps(completion: @escaping (Result<ApiOTAGpsInfo?, any Error>) -> Void) {
        guard let gpsInfo = delegateGps?.getAppGpsInfo() else {
            completion(.success(nil))
            return
        }
        let rs = ApiOTAGpsInfo(tcxoOffset: 0,
                               longitude: Double(gpsInfo.longitude),
                               latitude: Double(gpsInfo.latitude),
                               altitude: Double(gpsInfo.altitude))
        completion(.success(rs))
    }
    
    
}
